crates/matrix-sdk/src/client/login_builder.rs

```diff
@@ -0,0 +1,313 @@
+#![cfg_attr(not(target_arch = "wasm32"), deny(clippy::future_not_send))]
+
+#[cfg(all(feature = "sso-login", not(target_arch = "wasm32")))]
+use std::future::Future;
+
+use ruma::{
+    api::client::{session::login, uiaa::UserIdentifier},
+    assign,
+};
+use tracing::{info, instrument};
+
+use super::Client;
+use crate::{config::RequestConfig, Result};
+
+enum LoginMethod<'a> {
+    UserPassword { id: UserIdentifier<'a>, password: &'a str },
+    Token(&'a str),
+}
+
+impl<'a> LoginMethod<'a> {
+    fn id(&self) -> Option<UserIdentifier<'a>> {
+        match self {
+            LoginMethod::UserPassword { id, .. } => Some(id.clone()),
+            LoginMethod::Token(_) => None,
+        }
+    }
```

Comment:
Since we already have the lifetime `'a` here, nothing prevent us to return an `Option<&'a UserIdentifier<'a>>`? It would save a clone.

Thoughts?

Reply:
Not `&'a`, but just `&` works. I'll update accordingly.